src/scheduler.jl

```diff
@@ -89,9 +89,20 @@ function compute_dag(ctx, d::Thunk; options=SchedulerOptions())
     node_order = x -> -get(ord, x, 0)
     state = start_state(deps, node_order)
     # start off some tasks
+    # Note: worker_state may be different things for different contexts. Don't touch it out here!
     procs_state = assign_new_procs!(ctx, state, chan, node_order)
     @dbg timespan_end(ctx, :scheduler_init, 0, master)
 
+    # Check periodically for new workers in a parallel task so that we don't accidentally end up
```

Comment:
I don't think is necessary or desired; if there's the possibility that we could be waiting on `take!(chan)` for completed work that doesn't exist, then we should do a check that `!isempty(state.running)` prior to waiting. That would ensure at least one thunk is in-flight and will eventually return to us.

Also, currently this will run forever (even after the scheduler completes the DAG).

Reply:
Hmmm, it was intended as an optimization more than handling when chan is never populated.

In my use case which triggered the desire for this functionality, I have about 400 files which each take 2-30 minutes to process. Assume that the cluster scheduler starts by dispatching say 10 out of 400 jobs. If I'm really unlucky those 10 workers each get to work on a 30 minute file and even if I get more workers in the meantime no new tasks will be started by the scheduler until one of those 10 workers are done with their current task since `take!` blocks the thread. Note that it might take way more than 30 minutes to get all 400 jobs up and running, maybe several days if the cluster is heaviliy loaded.

I don't love that the loop is littered with locks now so I can try something else if that is the problem, but I think the use case is there. Or have I misunderstood something?

Reply:
I put the same condition in the asynch task as in the main scheduling loop, so it should stop when there is nothing to compute and nothing running. Will double check that now just in case...

Reply:
> Also, currently this will run forever (even after the scheduler completes the DAG).

> I put the same condition in the asynch task as in the main scheduling loop, so it should stop when there is nothing to compute and nothing running. Will double check that now just in case...

Sorry, I misread the implementation, you're right!

> In my use case which triggered the desire for this functionality, I have about 400 files which each take 2-30 minutes to process. Assume that the cluster scheduler starts by dispatching say 10 out of 400 jobs. If I'm really unlucky those 10 workers each get to work on a 30 minute file and even if I get more workers in the meantime no new tasks will be started by the scheduler until one of those 10 workers are done with their current task since take! blocks the thread. Note that it might take way more than 30 minutes to get all 400 jobs up and running, maybe several days if the cluster is heaviliy loaded.

This is an interesting case that I had considered but never fixed in the scheduler. I think the long-term solution is to split off the receiving of completed work from the channel into an `@async` or `@spawn` task, and having one or more separate tasks working on sending work to the workers. This seems to accomplish that, so I'm on board with your implementation.

Reply:
Awesome! I will take another pass to see if I can reduce the amount of moving parts. Afaik, I need to lock down all cases of pop_with_affinity! to prevent that the same task is scheduled twice, right? I see now that I have missed doing this for the fault handling.

I'm thinking that the only place where it is necessary to check for new workers is when waiting for `take!(chan)`, so perhaps either keep the lock locked everywhere in the loop except there or spawn the task right before waiting if the channel is empty and then stop it (with a flag) right after something popped out of the channel. 

Thoughts on this?